### PR TITLE
Fix BC-break on delete without alias DQL

### PR DESCRIPTION
**Pull request by lcobucci**
As mentioned in #7053, this fixes the issue without breaking other tests.

### CODE REVIEW

File: lib/Doctrine/ORM/Query/Parser.php [hunk 1]
```diff
@@ -1280,7 +1281,15 @@ public function DeleteClause()
             $this->match(Lexer::T_AS);
         }
 
-        $aliasIdentificationVariable = $this->AliasIdentificationVariable();
+        try {
```
Comment (lcobucci) [inline lib/Doctrine/ORM/Query/Parser.php]: Maybe we should a have a comment here to explain the 💩 and specially mention that we should raise a deprecation notice in `v2.7.0`

File: lib/Doctrine/ORM/Query/Parser.php [hunk 2]
```diff
@@ -304,21 +304,24 @@ public function match($token)
         $lookaheadType = $this->lexer->lookahead['type'];
 
         // Short-circuit on first condition, usually types match
-        if ($lookaheadType !== $token) {
-            // If parameter is not identifier (1-99) must be exact match
-            if ($token < Lexer::T_IDENTIFIER) {
-                $this->syntaxError($this->lexer->getLiteral($token));
-            }
+        if ($lookaheadType === $token) {
```
Comment (lcobucci) [inline lib/Doctrine/ORM/Query/Parser.php]: I can drop this commit if you prefer, it was just to make things a bit more sane.

**Comment by lcobucci** [comment]
@Ocramius @guilhermeblanco I was checking JPA 2.2 specs and maybe we should rather update the grammar for bulk delete:

> delete_statement ::= delete_clause [where_clause]
delete_clause ::= DELETE FROM entity_name [[AS] identification_variable]

**Source:** http://download.oracle.com/otndocs/jcp/persistence-2_2-mrel-spec/index.html (4.10 Bulk Update and Delete Operations)

File: lib/Doctrine/ORM/Query/Parser.php [hunk 3]
```diff
@@ -1280,7 +1284,15 @@ public function DeleteClause()
             $this->match(Lexer::T_AS);
         }
 
-        $aliasIdentificationVariable = $this->AliasIdentificationVariable();
+        try {
```
Comment (carnage) [inline lib/Doctrine/ORM/Query/Parser.php]: Could you use $this->lexer->isNextToken for this instead of trying to handle the exception?

Comment (lcobucci) [reply lib/Doctrine/ORM/Query/Parser.php]: Yes indeed, updating

Comment (lcobucci) [reply lib/Doctrine/ORM/Query/Parser.php]: @carnage thanks... it was a way too gambiarra before HAHAHA

File: lib/Doctrine/ORM/Query/Parser.php (hunk 2 above)
Comment (Ocramius) [reply lib/Doctrine/ORM/Query/Parser.php]: Works as-is

File: lib/Doctrine/ORM/Query/Parser.php [hunk 4]
```diff
@@ -1280,7 +1284,9 @@ public function DeleteClause()
             $this->match(Lexer::T_AS);
         }
 
-        $aliasIdentificationVariable = $this->AliasIdentificationVariable();
+        $aliasIdentificationVariable = $this->lexer->isNextToken(Lexer::T_IDENTIFIER)
+            ? $this->AliasIdentificationVariable()
+            : 'alias_should_have_been_set';
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Query/Parser.php]: We should likely raise a deprecation notice here, to clarify that this isn't supported DQL (according to BNF)

File: tests/Doctrine/Tests/ORM/Query/DeleteSqlGenerationTest.php [hunk 5]
```diff
@@ -36,6 +36,14 @@ public function assertSqlGeneration($dqlToBeTested, $sqlToBeConfirmed)
         }
     }
 
+    public function testSupportsDeleteWithoutWhereAndAlias() : void
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Query/DeleteSqlGenerationTest.php]: Please add an `@group` to this

**Review by Ocramius — CHANGES_REQUESTED** [review]
Another location to check is `Doctrine\ORM\QueryBuilder#delete()`, which allows for an optional alias. We should probably raise a deprecation notice there too...

File: lib/Doctrine/ORM/Query/Parser.php (hunk 2 above)
Comment (lcobucci) [reply lib/Doctrine/ORM/Query/Parser.php]: It works indeed, it's just not so good as it could be... I can send another PR to make you happier 😂

**Comment by lcobucci** [comment]
@Ocramius and what's your opinion regarding changing the grammar?

**Comment by Ocramius** [comment]
> what's your opinion regarding changing the grammar?

Only after we move to https://github.com/hoaproject/Compiler

Documenting grammar BC breaks becomes trivial after that. No changes before that, IMO.

File: lib/Doctrine/ORM/Query/Parser.php (hunk 4 above)
Comment (lcobucci) [reply lib/Doctrine/ORM/Query/Parser.php]: Didn't we say that deprecation notices were going to be added in `v2.7` only?

Comment (Ocramius) [reply lib/Doctrine/ORM/Query/Parser.php]: Works for me as long as you please create an issue tracking this specific deprecation 👍

Comment (lcobucci) [reply lib/Doctrine/ORM/Query/Parser.php]: Sure, I was going to do that anyway 😄

**Comment by Ocramius** [comment]
This has been forward-ported to `master` in doctrine/doctrine2@92445d095e8b11a98507438215b60cc6bab500a7

**Pull request merged**